### PR TITLE
feat(unit-price): Compute amount details for graduated charge model

### PR DESCRIPTION
**Pull request by rsempe**
## Roadmap Task

👉  [https://getlago.canny.io/feature-requests/p/display-unit-price-fee-breakdown](https://getlago.canny.io/feature-requests/p/display-unit-price-fee-breakdown)

## Context

Invoices are not compliant as we’re not displaying unit prices of item on them.

The goal is to be compliant by adding unit price on invoice payload and pdf.
- Subscription invoices
- Charge (pay in advance) invoices

## Description

The goal of this PR is to compute amount details for the graduated charge model.

Here is how it will be presented on the invoice:
Screenshot 2023-11-16 at 21 01 58

### CODE REVIEW

File: app/services/charges/charge_models/graduated_service.rb [hunk 1]
```diff
@@ -9,37 +9,24 @@ def ranges
...
-          break result_amount if range[:to_value].nil? || range[:to_value] >= units
-
-          result_amount
-        end
+      def amount_details
+        {
+          graduated_ranges: ranges.each_with_object([]) do |range, amounts|
+            amounts << Charges::AmountDetails::RangeGraduatedService.call(range:, total_units: units)
+            break amounts if range[:to_value].nil? || range[:to_value] >= units
+          end
```
Comment (jdenquin) [inline app/services/charges/charge_models/graduated_service.rb:17]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/services/charges/amount_details/range_graduated_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  let(:range) do
+    {
+      from_value: 0,
+      to_value: 10,
+      per_unit_amount: '10',
+      flat_amount: '2',
+    }
+  end
+
+  it "returns expected amount details" do
```
Comment (jdenquin) [inline spec/services/charges/amount_details/range_graduated_service_spec.rb:17]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/services/charges/amount_details/range_graduated_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    expect(service.call).to eq(
+      {
+        from_value: 0,
+        to_value: 10,
+        flat_amount: 2,
+        unit_amount: 10,
+        units: 10,
+        total_amount: 100,
+        total_with_flat_amount: 102,
+      }
```
Comment (jdenquin) [inline spec/services/charges/amount_details/range_graduated_service_spec.rb:27]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/charges/amount_details/range_graduated_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    )
+  end
+
+  context "when total units <= range to_value" do
```
Comment (jdenquin) [inline spec/services/charges/amount_details/range_graduated_service_spec.rb:31]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/services/charges/amount_details/range_graduated_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    let(:range) do
+      {
+        from_value: 11,
+        to_value: 20,
+        per_unit_amount: '8',
+        flat_amount: '1',
+      }
+    end
+
+    it "returns expected amount details" do
```
Comment (jdenquin) [inline spec/services/charges/amount_details/range_graduated_service_spec.rb:41]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/services/charges/amount_details/range_graduated_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      expect(service.call).to eq(
+        {
+          from_value: 11,
+          to_value: 20,
+          flat_amount: 1,
+          unit_amount: 8,
+          units: 5,
+          total_amount: 40,
+          total_with_flat_amount: 41,
+        }
```
Comment (jdenquin) [inline spec/services/charges/amount_details/range_graduated_service_spec.rb:51]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 7]
```diff
@@ -48,48 +48,176 @@
...
+          graduated_ranges: [
+            {
+              flat_amount: 2,
+              from_value: 0,
+              to_value: 10,
+              total_amount: 100,
+              total_with_flat_amount: 102,
+              unit_amount: 10,
+              units: 10,
+            }, {
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:137]: Layout/MultilineArrayLineBreaks: Each item in a multi-line array must start on a separate line.

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 8]
```diff
@@ -48,48 +48,176 @@
...
+              flat_amount: 3,
+              from_value: 11,
+              to_value: 20,
+              total_amount: 5,
+              total_with_flat_amount: 8,
+              unit_amount: 5,
+              units: 1,
+            },
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:145]: Style/TrailingCommaInArrayLiteral: Avoid comma after the last item of an array, unless each item is on its own line. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 9]
```diff
@@ -48,48 +48,176 @@
...
+          graduated_ranges: [
+            {
+              flat_amount: 2,
+              from_value: 0,
+              to_value: 10,
+              total_amount: 100,
+              total_with_flat_amount: 102,
+              unit_amount: 10,
+              units: 10,
+            }, {
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:169]: Layout/MultilineArrayLineBreaks: Each item in a multi-line array must start on a separate line.

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 10]
```diff
@@ -48,48 +48,176 @@
...
+              flat_amount: 3,
+              from_value: 11,
+              to_value: 20,
+              total_amount: 10,
+              total_with_flat_amount: 13,
+              unit_amount: 5,
+              units: 2,
+            },
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:177]: Style/TrailingCommaInArrayLiteral: Avoid comma after the last item of an array, unless each item is on its own line. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 11]
```diff
@@ -48,48 +48,176 @@
...
+          graduated_ranges: [
+            {
+              flat_amount: 2,
+              from_value: 0,
+              to_value: 10,
+              total_amount: 100,
+              total_with_flat_amount: 102,
+              unit_amount: 10,
+              units: 10,
+            }, {
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:201]: Layout/MultilineArrayLineBreaks: Each item in a multi-line array must start on a separate line.

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 12]
```diff
@@ -48,48 +48,180 @@
...
+            },
+            {
+              flat_amount: 3,
+              from_value: 11,
+              to_value: 20,
+              total_amount: 5,
+              total_with_flat_amount: 8,
+              unit_amount: 5,
+              units: 1,
+            }
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:146]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 13]
```diff
@@ -48,48 +48,180 @@
...
+            },
+            {
+              flat_amount: 3,
+              from_value: 11,
+              to_value: 20,
+              total_amount: 10,
+              total_with_flat_amount: 13,
+              unit_amount: 5,
+              units: 2,
+            }
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:179]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 14]
```diff
@@ -48,48 +48,180 @@
...
+            },
+            {
+              flat_amount: 3,
+              from_value: 11,
+              to_value: 20,
+              total_amount: 50,
+              total_with_flat_amount: 53,
+              unit_amount: 5,
+              units: 10,
+            },
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:212]: Layout/TrailingWhitespace: Trailing whitespace detected. (https://rubystyle.guide#no-trailing-whitespace)

File: spec/services/charges/charge_models/graduated_service_spec.rb [hunk 15]
```diff
@@ -48,48 +48,180 @@
...
+            }, 
+            {
+              flat_amount: 3,
+              from_value: 21,
+              to_value: nil,
+              total_amount: 5,
+              total_with_flat_amount: 8,
+              unit_amount: 5,
+              units: 1,
+            }
```
Comment (jdenquin) [inline spec/services/charges/charge_models/graduated_service_spec.rb:221]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**